### DISCUSSION TITLE
[question] Does DART support ray collisions?

### DISCUSSION THREAD

**Discussion by romulogcerqueira** [dartsim/dart#1803] [Q&A] [answered]
Hi,

I'm interested in confirming if DART supports ray collisions. I could not find any mention in the DART documentation, just a few documentation code snippets at [OdeCollisionDetector.hpp](https://github.com/dartsim/dart/blob/main/dart/collision/ode/OdeCollisionDetector.hpp#L50) and [CollisionDetector.cpp](https://github.com/dartsim/dart/blob/4935fb21c8eb25b0cc78686bb4f3e9595dc9539f/dart/collision/CollisionDetector.cpp#L67).

Thanks in advance.

**Comment by jslee02** [comment]
Yes, DART does support ray collisions via the Bullet backend. You may find the [`CollisionDetector::raycast()`](https://dartsim.github.io/dart/v6.13.2/d6/de8/classdart_1_1collision_1_1CollisionDetector.html#ab2e0b5b72bb4a71a9c81c19580b5cd06) and [`CollisionGroup::raycast()`](https://dartsim.github.io/dart/v6.13.2/d2/d7a/classdart_1_1collision_1_1CollisionGroup.html#a9d7c097d26f0c30b7794fab9793dfb2d) APIs useful. Although we don't have a C++ example at the moment, we do have a [Python example](https://github.com/dartsim/dart/blob/4935fb21c8eb25b0cc78686bb4f3e9595dc9539f/python/tests/unit/collision/test_collision.py#L198) that could provide some idea on how to use these C++ APIs.

**Comment by romulogcerqueira** [comment]
Thanks for your reply, @jslee02.

My question was related to the missing ray collision feature in [gz-sim](https://github.com/gazebosim/gz-sim), where DART is the default physics engine. There is an [open issue](https://github.com/gazebosim/gz-sensors/issues/26) for that.

Do you have any recommendations for using the Bullet backend in [`gz-physics`](https://github.com/gazebosim/gz-physics)?

Alternatively, can we use [`ConeShape`](https://github.com/dartsim/dart/blob/main/dart/dynamics/ConeShape.hpp) to mimic the ray collision? Does DART present the contact points/closest points for a collision between cones and 3D arbitrary in the scene?

**Comment by jslee02** [comment, marked as answer]
Yes. DART supports ray queries, but the supported path depends on the DART version and collision backend.

For DART 6.x, raycast support is available through the Bullet collision detector backend. The relevant APIs are `CollisionDetector::raycast()` and `CollisionGroup::raycast()`, with `RaycastOption` and `RaycastResult` for closest/all-hit behavior and hit details. In Gazebo/gz-sim, this means the practical path is to configure the DART/Bullet collision detector through the gz-physics/SDF integration; the exact sensor feature exposure is handled on the Gazebo side.

For DART 7/main, DART's native collision backend also supports raycast, so Bullet is no longer required for DART's own raycast API on current `main`.

Using `ConeShape` can approximate a finite-volume sensor shape, but it is not the same as a zero-width ray query. It will produce ordinary collision/contact results for a cone-shaped volume; for ray-style sensor intersection, prefer the raycast API.
